src/utils/errors.ts

```diff
@@ -0,0 +1,54 @@
+import z from 'zod';
+
+import { ErrorResponse, ErrorResponseErrorEnum } from '../types';
+
+const ErrorResponseSchema: z.ZodType<ErrorResponse> = z.object({
+  code: z.number(),
+  detail: z.optional(z.string()),
+  error: z.nativeEnum(ErrorResponseErrorEnum),
+});
+
+export class SpeechmaticsResponseError extends Error {
+  response: ErrorResponse;
+
+  constructor(errorResponse: unknown) {
+    const parse = ErrorResponseSchema.safeParse(errorResponse);
+    if (parse.success) {
+      super(parse.data.error);
+      this.response = parse.data;
+    } else {
+      throw new SpeechmaticsInternalError('Unexpected response');
+    }
+  }
+}
+
+export const InternalErrorEnum = {
+  ProcessUndefined: 'process is undefined - are you running in node?',
+  WindowUndefined: 'window is undefined - are you running in a browser?',
+  ApiKeyUndefined: 'Error: apiKey is undefined',
+  FetchSLT: 'Error fetching short lived token',
+  UnsuportedEnvironment: 'Unsupported environment',
+  UnexpectedMessage: 'Unexpected message',
+  UnexpectedResponse: 'Unexpected response',
+  TypeError: 'Unexpected type',
+} as const;
+
+export type InternalErrorEnum =
+  typeof InternalErrorEnum[keyof typeof InternalErrorEnum];
+
+export class SpeechmaticsInternalError extends Error {
+  error: InternalErrorEnum;
+  detail?: string;
+  cause?: unknown;
+
+  constructor(error: InternalErrorEnum, detail?: string, cause?: unknown) {
+    super();
+    this.error = error;
+    this.detail = detail;
+    this.cause = cause;
+  }
+}
+
+export type SpeechmaticsError =
+  | SpeechmaticsInternalError
+  | SpeechmaticsResponseError;
```

Comment:
As a consumer of these errors I should be able to quickly distinguish between them.

At the moment we have an enum for each, but on the internal error the enum is at `.error` and for the response it's at `.response.error`.

May need to review this for consistency between the two types.

Reply:
Maybe for the response error we also include the enum at `.error`. This would allow us to also incorporate the UnexpectedResponse error here, rather than using an InternalError.

Reply:
I think for now it's probably OK given that we export all of the variants (of both types) at the top level. I don't imagine many users will be checking specifically whether it's an `Internal` or `Response` error, so we can let the type guards apply to the individual variants